### PR TITLE
Unroll loop for Utf8Parser unsigned integer parsers

### PR DESCRIPTION
**Pull request by AtsushiKan**
This optimizes the Byte/UInt16/UInt32 parsers in the same
way that their signed counterparts already were.

Also, tuned up the signed counterparts before copy-pasting them:

- Just subtract the '0' once rather than doing it once in IsDigit()
  and then again to add to the accumulator.

- Eliminate the pre-overflow checks in the case where
  the accumulator is already a larger sized int than the value
  being parsed. The accumulator will never overflow as long
  we're under the digit count limit for the type.

- Fix the Int32 parser's potential overflow comparison
  from "int.MaxValue/10 - 1" to "int.MaxValue/10" (the older
  value works with an unsigned accumulator but in a "by accident" type
  of way.)

Perf improvement:

  +2.6% Int32  (int.MaxValue)
  +7.0% UInt32 (uint.MaxValue)

### CODE REVIEW

**Comment by ahsonkhan** [comment]
> Just subtract the '0' once rather than doing it once in IsDigit()
and then again to add to the accumulator.

The disassembly now is ~7 bytes larger (i.e. 2 instructions) and it looks like it is actually a bit slower (rather than faster). cc @jkotas 
image


Test   Name | Metric | Before | After | Ratio
-- | -- | -- | -- | --
ParseInt32(text:   "+2") | Duration | 0.074 | 0.079 | 1.066
ParseInt32(text:   "+21") | Duration | 0.081 | 0.085 | 1.056
ParseInt32(text:   "+214") | Duration | 0.084 | 0.089 | 1.059
ParseInt32(text:   "+2147") | Duration | 0.085 | 0.102 | 1.193
ParseInt32(text:   "+21474") | Duration | 0.107 | 0.133 | 1.243
ParseInt32(text:   "+214748") | Duration | 0.101 | 0.120 | 1.191
ParseInt32(text:   "+2147483") | Duration | 0.108 | 0.110 | 1.016
ParseInt32(text:   "+21474836") | Duration | 0.117 | 0.130 | 1.108
ParseInt32(text:   "+214748364") | Duration | 0.119 | 0.128 | 1.073
ParseInt32(text:   "+2147483647") | Duration | 0.159 | 0.165 | 1.036
ParseInt32(text:   "+2147483647abcdefghijklmnop") | Duration | 0.147 | 0.159 | 1.079
ParseInt32(text:   "+214748364abcdefghijklmnop") | Duration | 0.123 | 0.132 | 1.074
ParseInt32(text:   "+21474836abcdefghijklmnop") | Duration | 0.125 | 0.124 | 0.998
ParseInt32(text:   "+2147483abcdefghijklmnop") | Duration | 0.113 | 0.117 | 1.037
ParseInt32(text:   "+214748abcdefghijklmnop") | Duration | 0.109 | 0.108 | 0.996
ParseInt32(text:   "+21474abcdefghijklmnop") | Duration | 0.097 | 0.107 | 1.094
ParseInt32(text:   "+2147abcdefghijklmnop") | Duration | 0.094 | 0.098 | 1.047
ParseInt32(text:   "+214abcdefghijklmnop") | Duration | 0.088 | 0.092 | 1.048
ParseInt32(text:   "+21abcdefghijklmnop") | Duration | 0.078 | 0.090 | 1.158
ParseInt32(text:   "+2abcdefghijklmnop") | Duration | 0.079 | 0.083 | 1.043
ParseInt32(text:   "0") | Duration | 0.072 | 0.086 | 1.203
ParseInt32(text:   "000000000000000000001235abcdfg") | Duration | 0.277 | 0.224 | 0.808
ParseInt32(text:   "107374182") | Duration | 0.128 | 0.137 | 1.072
ParseInt32(text:   "2") | Duration | 0.079 | 0.082 | 1.049
ParseInt32(text:   "-2") | Duration | 0.079 | 0.081 | 1.019
ParseInt32(text:   "21") | Duration | 0.084 | 0.089 | 1.069
ParseInt32(text:   "-21") | Duration | 0.079 | 0.086 | 1.084
ParseInt32(text:   "214") | Duration | 0.119 | 0.092 | 0.770
ParseInt32(text:   "-214") | Duration | 0.087 | 0.092 | 1.056
ParseInt32(text:   "2147") | Duration | 0.095 | 0.101 | 1.068
ParseInt32(text:   "-2147") | Duration | 0.092 | 0.096 | 1.045
ParseInt32(text:   "21474") | Duration | 0.101 | 0.101 | 1.003
ParseInt32(text:   "-21474") | Duration | 0.099 | 0.107 | 1.085
ParseInt32(text:   "214748") | Duration | 0.107 | 0.111 | 1.038
ParseInt32(text:   "-214748") | Duration | 0.107 | 0.136 | 1.276
ParseInt32(text:   "2147483") | Duration | 0.117 | 0.114 | 0.974
ParseInt32(text:   "-2147483") | Duration | 0.117 | 0.111 | 0.949
ParseInt32(text:   "21474836") | Duration | 0.122 | 0.134 | 1.104
ParseInt32(text:   "-21474836") | Duration | 0.109 | 0.158 | 1.452
ParseInt32(text:   "214748364") | Duration | 0.124 | 0.130 | 1.046
ParseInt32(text:   "-214748364") | Duration | 0.166 | 0.132 | 0.794
ParseInt32(text:   "2147483647") | Duration | 0.149 | 0.146 | 0.978
ParseInt32(text:   "-2147483648") | Duration | 0.144 | 0.145 | 1.013
ParseInt32(text:   "214748364abcdefghijklmnop") | Duration | 0.122 | 0.130 | 1.070
ParseInt32(text:   "-214748364abcdefghijklmnop") | Duration | 0.137 | 0.132 | 0.960
ParseInt32(text:   "21474836abcdefghijklmnop") | Duration | 0.118 | 0.158 | 1.338
ParseInt32(text:   "-21474836abcdefghijklmnop") | Duration | 0.117 | 0.121 | 1.037
ParseInt32(text:   "2147483abcdefghijklmnop") | Duration | 0.117 | 0.115 | 0.987
ParseInt32(text:   "-2147483abcdefghijklmnop") | Duration | 0.116 | 0.121 | 1.039
ParseInt32(text:   "214748abcdefghijklmnop") | Duration | 0.110 | 0.111 | 1.004
ParseInt32(text:   "-214748abcdefghijklmnop") | Duration | 0.103 | 0.113 | 1.089
ParseInt32(text:   "21474abcdefghijklmnop") | Duration | 0.104 | 0.107 | 1.026
ParseInt32(text:   "-21474abcdefghijklmnop") | Duration | 0.100 | 0.102 | 1.014
ParseInt32(text:   "2147abcdefghijklmnop") | Duration | 0.095 | 0.100 | 1.059
ParseInt32(text:   "-2147abcdefghijklmnop") | Duration | 0.094 | 0.100 | 1.070
ParseInt32(text:   "214abcdefghijklmnop") | Duration | 0.095 | 0.093 | 0.988
ParseInt32(text:   "-214abcdefghijklmnop") | Duration | 0.086 | 0.094 | 1.090
ParseInt32(text:   "21abcdefghijklmnop") | Duration | 0.088 | 0.093 | 1.054
ParseInt32(text:   "-21abcdefghijklmnop") | Duration | 0.081 | 0.084 | 1.038
ParseInt32(text:   "2abcdefghijklmnop") | Duration | 0.081 | 0.083 | 1.024
ParseInt32(text:   "-2abcdefghijklmnop") | Duration | 0.078 | 0.083 | 1.063
ParseInt32_VariableLength | Duration | 0.154 | 0.147 | 0.954



Test   Name | Metric | Before | After | Ratio
-- | -- | -- | -- | --
ParseInt32(text:   "+2") | Instructions Retired | 630000 | 610000 | 0.968
ParseInt32(text:   "+21") | Instructions Retired | 690000 | 800000 | 1.159
ParseInt32(text:   "+214") | Instructions Retired | 800000 | 830000 | 1.038
ParseInt32(text:   "+2147") | Instructions Retired | 920000 | 950000 | 1.033
ParseInt32(text:   "+21474") | Instructions Retired | 1040000 | 1060000 | 1.019
ParseInt32(text:   "+214748") | Instructions Retired | 1140000 | 1180000 | 1.035
ParseInt32(text:   "+2147483") | Instructions Retired | 1250000 | 1280000 | 1.024
ParseInt32(text:   "+21474836") | Instructions Retired | 1340000 | 1380000 | 1.030
ParseInt32(text:   "+214748364") | Instructions Retired | 1460000 | 1530000 | 1.048
ParseInt32(text:   "+2147483647") | Instructions Retired | 1710000 | 1730000 | 1.012
ParseInt32(text:   "+2147483647abcdefghijklmnop") | Instructions Retired | 1890000 | 1840000 | 0.974
ParseInt32(text:   "+214748364abcdefghijklmnop") | Instructions Retired | 1550000 | 1540000 | 0.994
ParseInt32(text:   "+21474836abcdefghijklmnop") | Instructions Retired | 1470000 | 1470000 | 1.000
ParseInt32(text:   "+2147483abcdefghijklmnop") | Instructions Retired | 1320000 | 1340000 | 1.015
ParseInt32(text:   "+214748abcdefghijklmnop") | Instructions Retired | 1200000 | 1260000 | 1.050
ParseInt32(text:   "+21474abcdefghijklmnop") | Instructions Retired | 1100000 | 1120000 | 1.018
ParseInt32(text:   "+2147abcdefghijklmnop") | Instructions Retired | 1000000 | 1010000 | 1.010
ParseInt32(text:   "+214abcdefghijklmnop") | Instructions Retired | 880000 | 880000 | 1.000
ParseInt32(text:   "+21abcdefghijklmnop") | Instructions Retired | 680000 | 860000 | 1.265
ParseInt32(text:   "+2abcdefghijklmnop") | Instructions Retired | 680000 | 700000 | 1.029
ParseInt32(text:   "0") | Instructions Retired | 550000 | 560000 | 1.018
ParseInt32(text:   "000000000000000000001235abcdfg") | Instructions Retired | 2500000 | 2470000 | 0.988
ParseInt32(text:   "107374182") | Instructions Retired | 1460000 | 1450000 | 0.993
ParseInt32(text:   "2") | Instructions Retired | 570000 | 560000 | 0.982
ParseInt32(text:   "-2") | Instructions Retired | 580000 | 610000 | 1.052
ParseInt32(text:   "21") | Instructions Retired | 650000 | 700000 | 1.077
ParseInt32(text:   "-21") | Instructions Retired | 660000 | 770000 | 1.167
ParseInt32(text:   "214") | Instructions Retired | 790000 | 810000 | 1.025
ParseInt32(text:   "-214") | Instructions Retired | 880000 | 880000 | 1.000
ParseInt32(text:   "2147") | Instructions Retired | 880000 | 920000 | 1.045
ParseInt32(text:   "-2147") | Instructions Retired | 910000 | 970000 | 1.066
ParseInt32(text:   "21474") | Instructions Retired | 990000 | 1020000 | 1.030
ParseInt32(text:   "-21474") | Instructions Retired | 1040000 | 1130000 | 1.087
ParseInt32(text:   "214748") | Instructions Retired | 1110000 | 1160000 | 1.045
ParseInt32(text:   "-214748") | Instructions Retired | 1160000 | 1170000 | 1.009
ParseInt32(text:   "2147483") | Instructions Retired | 1210000 | 1240000 | 1.025
ParseInt32(text:   "-2147483") | Instructions Retired | 1300000 | 1270000 | 0.977
ParseInt32(text:   "21474836") | Instructions Retired | 1300000 | 1340000 | 1.031
ParseInt32(text:   "-21474836") | Instructions Retired | 1350000 | 1400000 | 1.037
ParseInt32(text:   "214748364") | Instructions Retired | 1450000 | 1400000 | 0.966
ParseInt32(text:   "-214748364") | Instructions Retired | 1480000 | 1520000 | 1.027
ParseInt32(text:   "2147483647") | Instructions Retired | 1740000 | 1700000 | 0.977
ParseInt32(text:   "-2147483648") | Instructions Retired | 1760000 | 1760000 | 1.000
ParseInt32(text:   "214748364abcdefghijklmnop") | Instructions Retired | 1500000 | 1460000 | 0.973
ParseInt32(text:   "-214748364abcdefghijklmnop") | Instructions Retired | 1530000 | 1560000 | 1.020
ParseInt32(text:   "21474836abcdefghijklmnop") | Instructions Retired | 1380000 | 1370000 | 0.993
ParseInt32(text:   "-21474836abcdefghijklmnop") | Instructions Retired | 1430000 | 1440000 | 1.007
ParseInt32(text:   "2147483abcdefghijklmnop") | Instructions Retired | 1230000 | 1320000 | 1.073
ParseInt32(text:   "-2147483abcdefghijklmnop") | Instructions Retired | 1310000 | 1340000 | 1.023
ParseInt32(text:   "214748abcdefghijklmnop") | Instructions Retired | 1140000 | 1190000 | 1.044
ParseInt32(text:   "-214748abcdefghijklmnop") | Instructions Retired | 1190000 | 1220000 | 1.025
ParseInt32(text:   "21474abcdefghijklmnop") | Instructions Retired | 1050000 | 1050000 | 1.000
ParseInt32(text:   "-21474abcdefghijklmnop") | Instructions Retired | 1110000 | 1120000 | 1.009
ParseInt32(text:   "2147abcdefghijklmnop") | Instructions Retired | 960000 | 950000 | 0.990
ParseInt32(text:   "-2147abcdefghijklmnop") | Instructions Retired | 1000000 | 1000000 | 1.000
ParseInt32(text:   "214abcdefghijklmnop") | Instructions Retired | 790000 | 800000 | 1.013
ParseInt32(text:   "-214abcdefghijklmnop") | Instructions Retired | 890000 | 880000 | 0.989
ParseInt32(text:   "21abcdefghijklmnop") | Instructions Retired | 750000 | 740000 | 0.987
ParseInt32(text:   "-21abcdefghijklmnop") | Instructions Retired | 740000 | 730000 | 0.986
ParseInt32(text:   "2abcdefghijklmnop") | Instructions Retired | 650000 | 580000 | 0.892
ParseInt32(text:   "-2abcdefghijklmnop") | Instructions Retired | 710000 | 720000 | 1.014

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs [hunk 1]
```diff
@@ -203,122 +203,124 @@ private static bool TryParseInt32D(ReadOnlySpan<byte> text, out int value, out i
...
                     goto Done;
-                num = text[index];
-                if (!ParserHelpers.IsDigit(num))
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
                     goto Done;
-                if (answer > Int32.MaxValue / 10 + 1)
-                    goto FalseExit; // Overflow
-                answer = answer * 10 + num - '0';
+                if (answer > int.MaxValue / 10)
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: Is this change needed **only** because the accumulator is now a uint rather than an int?

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs [hunk 2]
```diff
@@ -13,64 +13,64 @@ private static bool TryParseSByteD(ReadOnlySpan<byte> text, out sbyte value, out
...
-                    goto FalseExit; // Overflow
-                answer = answer * 10 + num - '0';
+                answer = 10 * answer + digit;
 
                 // if sign < 0, (-1 * sign + 1) / 2 = 1
                 // else, (-1 * sign + 1) / 2 = 0
-                if ((uint)answer > (uint)SByte.MaxValue + (-1 * sign + 1) / 2)
+                if (answer > (uint)sbyte.MaxValue + (-1 * sign + 1) / 2)
                     goto FalseExit; // Overflow
                 index++;
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: nit: move `index++` above `answer = 10 * answer + digit` to be consistent with other changes (like in TryParseInt32D [here](https://github.com/dotnet/corefx/pull/25130/files#diff-5e90497df430f6541664f0a780acac46R317))

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs [hunk 3]
```diff
@@ -9,225 +9,276 @@ public static partial class Utf8Parser
...
+                // Potential overflow
+                if ((uint)index >= (uint)text.Length)
+                    goto Done;
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
+                    goto Done;
+                answer = 10 * answer + digit;
+                if (answer > ushort.MaxValue)
+                    goto FalseExit; // Overflow
+                index++;
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs]: same nit here and elsewhere. keep the two lines together everywhere:
```C#
index++;
answer = 10 * answer + digit;
```

**Comment by ahsonkhan** [comment]
> This optimizes the Byte/UInt16/UInt32 parsers in the same
way that their signed counterparts already were.

This looks good to me.

> Also, tuned up the signed counterparts before copy-pasting them

I am not sure if these changes help necessarily but I could be wrong.

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs (hunk 1 above)
Comment (AtsushiKan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: The accumulator changing to uint doesn't have a bearing on this. Either way, the 10 extra values that leaked through that check get caught as overflows by the second check even after multiplying by 10. But either way, what was the purpose of the "+ 1"?

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs (hunk 2 above)
Comment (AtsushiKan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: Moved.

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs (hunk 3 above)
Comment (AtsushiKan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs]: Moved.

**Comment by AtsushiKan** [comment]
Humphf - what is about uint that the makes the JIT do that? Since the results are ambiguous at best, in the interests of keeping changes isolated, I've gone back to the master version for the signed versions (though I'm still removing the unnecessary pre-overflow checks and keeping your ask about keeping the "index++" placement consistent.)

File: src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: I can't figure out why we needed the "+ 1". I tried several test cases around int.MaxValue/10 and they all work with or without the "+ 1".

```C#
//(string text, bool expectSuccess, int expectedValue, int expectedConsumed)
[InlineData("214748364", true, 214748364, 9)] // int.MaxValue/10
[InlineData("214748363", true, 214748363, 9)]
[InlineData("214748362", true, 214748362, 9)]
[InlineData("214748361", true, 214748361, 9)]
[InlineData("214748365", true, 214748365, 9)]
[InlineData("214748366", true, 214748366, 9)]
[InlineData("214748367", true, 214748367, 9)]
[InlineData("214748368", true, 214748368, 9)]
[InlineData("214748369", true, 214748369, 9)]
[InlineData("214748370", true, 214748370, 9)]
[InlineData("2147483648", false, 0, 0)]
[InlineData("2147483649", false, 0, 0)]
[InlineData("2147483650", false, 0, 0)]
[InlineData("2147483647", true, 2147483647, 10)]
[InlineData("2147483646", true, 2147483646, 10)]
[InlineData("2147483645", true, 2147483645, 10)]
[InlineData("2147483644", true, 2147483644, 10)]
[InlineData("2147483643", true, 2147483643, 10)]
[InlineData("2147483642", true, 2147483642, 10)]
[InlineData("2147483641", true, 2147483641, 10)]
[InlineData("2147483640", true, 2147483640, 10)]
[InlineData("2147483639", true, 2147483639, 10)]
[InlineData("-214748364", true, -214748364, 10)] // int.MinValue/10
[InlineData("-214748363", true, -214748363, 10)]
[InlineData("-214748362", true, -214748362, 10)]
[InlineData("-214748361", true, -214748361, 10)]
[InlineData("-214748365", true, -214748365, 10)]
[InlineData("-214748366", true, -214748366, 10)]
[InlineData("-214748367", true, -214748367, 10)]
[InlineData("-214748368", true, -214748368, 10)]
[InlineData("-214748369", true, -214748369, 10)]
[InlineData("-214748370", true, -214748370, 10)]
[InlineData("-2147483648", true, -2147483648, 11)]
[InlineData("-2147483647", true, -2147483647, 11)]
[InlineData("-2147483646", true, -2147483646, 11)]
[InlineData("-2147483645", true, -2147483645, 11)]
[InlineData("-2147483644", true, -2147483644, 11)]
[InlineData("-2147483643", true, -2147483643, 11)]
[InlineData("-2147483642", true, -2147483642, 11)]
[InlineData("-2147483641", true, -2147483641, 11)]
[InlineData("-2147483640", true, -2147483640, 11)]
[InlineData("-2147483639", true, -2147483639, 11)]
[InlineData("-2147483649", false, 0, 0)]
[InlineData("-2147483650", false, 0, 0)]

        [InlineData("21474836460")] // false, 0, 0
        [InlineData("21474836480")] // false, 0, 0
        [InlineData("21474836490")] // false, 0, 0
        [InlineData("21474836500")] // false, 0, 0
        [InlineData("21474836450")] // false, 0, 0
        [InlineData("21474836440")] // false, 0, 0
        [InlineData("21474836430")] // false, 0, 0
        [InlineData("21474836420")] // false, 0, 0
        [InlineData("21474836410")] // false, 0, 0
        [InlineData("21474836400")] // false, 0, 0
        [InlineData("21474836390")] // false, 0, 0
        [InlineData("-21474836460")] // false, 0, 0
        [InlineData("-21474836480")] // false, 0, 0
        [InlineData("-21474836490")] // false, 0, 0
        [InlineData("-21474836500")] // false, 0, 0
        [InlineData("-21474836450")] // false, 0, 0
        [InlineData("-21474836440")] // false, 0, 0
        [InlineData("-21474836430")] // false, 0, 0
        [InlineData("-21474836420")] // false, 0, 0
        [InlineData("-21474836410")] // false, 0, 0
        [InlineData("-21474836400")] // false, 0, 0
        [InlineData("-21474836390")] // false, 0, 0
```

It came from a suggestion from Jan. Any ideas, Jan?
https://github.com/dotnet/corefxlab/pull/1616#issuecomment-310790392
cc @jkotas

Comment (jkotas) [reply src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs]: > It came from a suggestion from Jan. Any ideas, Jan?

Looks like oversight.

**Comment by AtsushiKan** [comment]
@ahsonkhan - do you have any other concerns?

**Comment by ahsonkhan** [comment]
> do you have any other concerns?

The changes LGTM.

As an aside: I wonder if we can improve the performance of the Int64/UInt64 parsers. Unrolling the loop for such types (like Int32/etc) would result in a method body that is too large but maybe there is another approach.
